### PR TITLE
Fix that the controller did not use creator's token to trigger a Jenkins job

### PR DESCRIPTION
**Pull request by JohnNiang**
### What dose this PR?

1. Add `devops.kubesphere.io/creator` into annotations of PipelineRun
2. Trigger PipelineRun with specific token which is issued by `devops.kubesphere.io/creator`

### Why do we need it?

At present, if we trigger a Pipeline using non-admin account, we will get message `Started by admin` always.

### Steps to test

Docker images for test:

```bash
johnniang/devops-apiserver:dev-v3.2.1-rc.1-2debd81
johnniang/devops-controller:dev-v3.2.1-rc.1-2debd81
```
1. Create a Workspace, DevOps Project, Pipeline using admin account
1. Create another account tester and invite him/her into that Workspace and DevOps Project
1. Trigger that Pipeline using tester account
1. See the `Last Message` column

image

/kind bug
/milestone v3.2
/cc @kubesphere/sig-devops

### CODE REVIEW

File: cmd/controller/app/options/options.go [hunk 1]
```diff
@@ -21,6 +21,7 @@ import (
 	"strings"
 	"time"
 
+	"kubesphere.io/devops/pkg/apiserver/authentication/options"
```
Comment (LinuxSuRen) [inline cmd/controller/app/options/options.go:24]: controller and apiserver should be independent of each other. Please avoid import the package of apiserver into the controller.

File: controllers/jenkins/pipelinerun/pipelinerun_controller.go [hunk 2]
```diff
@@ -172,8 +178,16 @@ func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 		return ctrl.Result{RequeueAfter: 3 * time.Second}, nil
 	}
 
+	// get or create JenkinsCore if the PipelineRun has creator annotation
+	jenkinsCore, err := r.getOrCreateJenkinsCoreIfHasCreator(pipelineRunCopied.GetAnnotations())
+	if err != nil {
+		r.recorder.Eventf(pipelineRunCopied, corev1.EventTypeWarning, v1alpha3.TriggerFailed, "Failed to create Jenkins core for triggering PipelineRun %s, and error was %s", req.NamespacedName, err)
```
Comment (LinuxSuRen) [inline controllers/jenkins/pipelinerun/pipelinerun_controller.go:184]: It's not easy to understand what `Jenkins Core` is.

File: controllers/jenkins/pipelinerun/pipelinerun_controller.go [hunk 3]
```diff
@@ -309,6 +323,24 @@ func (r *Reconciler) makePipelineRunOrphan(ctx context.Context, pr *v1alpha3.Pip
 	return r.updateStatus(ctx, &pr.Status, client.ObjectKey{Namespace: pr.Namespace, Name: pr.Name})
 }
 
+func (r *Reconciler) getOrCreateJenkinsCoreIfHasCreator(annotations map[string]string) (*core.JenkinsCore, error) {
```
Comment (LinuxSuRen) [inline controllers/jenkins/pipelinerun/pipelinerun_controller.go:326]: No matter if there is a creator or not, this function always returns an object. So, I suggest changing the function name to `getOrCreateJenkinsCore`

File: controllers/jenkins/pipelinerun/pipelinerun_controller.go [hunk 4]
```diff
@@ -309,6 +323,24 @@ func (r *Reconciler) makePipelineRunOrphan(ctx context.Context, pr *v1alpha3.Pip
...
+	creator, ok := annotations[v1alpha3.PipelineRunCreatorAnnoKey]
+	if !ok || creator == "" {
+		return &r.JenkinsCore, nil
+	}
+	// create a new JenkinsCore for current creator
+	accessToken, err := r.TokenIssuer.IssueTo(&user.DefaultInfo{Name: creator}, token.AccessToken, tokenExpireIn)
+	if err != nil {
+		return nil, fmt.Errorf("failed to issue access token for creator %s, error = %v", creator, err)
```
Comment (LinuxSuRen) [inline controllers/jenkins/pipelinerun/pipelinerun_controller.go:334]: Please try to keep using the same style at least for your own code lines.

`error = %v`

[line 184](https://github.com/kubesphere/ks-devops/pull/384/files#diff-e19db2bfc951270073311530a9c79cb225bad3986a778c4decefd366e853880bR184) `error was %s`

**Review by LinuxSuRen — CHANGES_REQUESTED** [review]

File: cmd/controller/app/options/options.go (hunk 1 above)
Comment (JohnNiang) [reply cmd/controller/app/options/options.go]: Done. Please see https://github.com/kubesphere/ks-devops/pull/384/commits/0e88c48d491c59e25df126da6704cb649ed415c5.

File: cmd/controller/app/controllers.go [hunk 5]
```diff
@@ -17,6 +17,7 @@ limitations under the License.
 package app
 
 import (
+	"kubesphere.io/devops/pkg/apiserver/authentication/token"
```
Comment (LinuxSuRen) [inline cmd/controller/app/controllers.go:20]: You are still using golang packages from apiserver in the code base of a controller.

**Review by LinuxSuRen — CHANGES_REQUESTED** [review]

File: cmd/controller/app/controllers.go (hunk 5 above)
Comment (JohnNiang) [reply cmd/controller/app/controllers.go]: Done. Please see https://github.com/kubesphere/ks-devops/pull/384/commits/2debd818f129aba4400ae4dd5d61591dcef57e37.

**Comment by JohnNiang** [comment]
I've run `make test` in my local environment many times, and it always passes.

image

I'm going to re-run the workflow by closing and reopening this PR.

**Comment by LinuxSuRen** [comment]
hi @JohnNiang , The real problem is that the controller didn't use the correct user to trigger a Jenkins job.

`showing wrong message "started by admin" while triggering by non-admin user` describes a UI issue.

So, I suggest you provide a more accurate title.

**Comment by JohnNiang** [comment]
> hi @JohnNiang , The real problem is that the controller didn't use the correct user to trigger a Jenkins job.
> 
> `showing wrong message "started by admin" while triggering by non-admin user` describes a UI issue.
> 
> So, I suggest you provide a more accurate title.

Thanks for your suggestion!

/retitle Enable the controller to trigger a Jenkins job with the token of the creator

**Review by LinuxSuRen — COMMENTED** [review]
I have tested it manually. It works well.
/lgtm
/approve

**Pull request merged**